### ISSUE TITLE
Access to Server Admin Panels

### ISSUE THREAD

**Issue by KoleBunny** [sbpp/sourcebans-pp#208]
Okay so i'm new to typing out exploits, but it's a pretty simple one. On all old SourceBans versions if you create an admin account with a steam account, it leaves it so you can just enter an admin name and login to the admin panel without a password or with their steam account. So if you're a server owner running it, please switch to the new 1.5.5-Dev. 
EDIT:
I was assuming this was the problem which I shouldn't have, issue is based of how keeping an account locked worked in old Sourcebans. Ignore everything said above this. 
![f50cfc63a98a0db4915c013fa134b598](https://cloud.githubusercontent.com/assets/11532552/22448408/339bf274-e727-11e6-8376-a344e7e656ee.png)

**Comment by eNNercY1337** [comment]
How about you set a password when generating users? Stupidity is not an exploit.

**Comment by KoleBunny** [comment]
Well you'd assume if you'd login through steam, said account would only be able to be accessed by your steam account, not the login/pass.

**Comment by RonMelkhior** [comment]
Eh no, the Steam login system was added ONTO to login system. Not the other way around.

**Comment by KoleBunny** [comment]
And even if it isn't really an exploit, server admins need to update, or remove perms from all accounts that aren't theirs. It seems like a common thing since i've used it on 2 servers. 
Video proof: https://youtu.be/iFZDutZKIf8
I just want to make server owners aware that this is a thing.

**Comment by RonMelkhior** [comment]
Blame the server owners for not being able to setup a web app properly.  
That'd be like blaming a working car for running over a person, instead of blaming the driver.

EDIT: I'm in no way saying SourceBans is perfect, but this one is just simply not an exploit, sure it might not need to be able to let people set no password as a password, but people need to be able to deal with simple stuff like this.

**Comment by KoleBunny** [comment]
I assumed when they set up their account with a steam account, said person thinks their account will only be able to be accessed by that steam account login. And it looks like a common thing, the new sourceban update though fixes that and only allows steam account logins, not making one on the sourcebans application right?
Like I said i'm not the best with server stuff this just feels weird that you can do this to servers.

**Comment by RonMelkhior** [comment]
> I assumed when they set up their account with a steam account

The only way to create an account is manually? And you need to setup your password there. lol

**Comment by KoleBunny** [comment]
Never used this application, I guess this is just a weird thing I ran into 2 servers with incompetent admins. Sorry for the post, should I remove it? Since it's really an an awareness to upgrade to the newer sourcebans that only let you login through steam (As I was told)

**Comment by RonMelkhior** [comment]
Yeah. I don't think issues should be used for this purpose. But you can let the maintainers be the judge of that.

**Comment by eNNercY1337** [comment]
Btw. @KoleBunny thanks for the sharing of the video, made my day. :grin:

**Comment by KoleBunny** [comment]
Yeah, I of course didn't want to fuck up any of there servers with rcon commands, or unban every steam acccount ever. I informed the server owners in this video about the problem, and I hope they decide to fix it.

**Comment by Munzy** [comment]
Back in the day with regular sourcebans I remember testing this. The account was effectively locked out, and I did this on purpose. Later down the road I guess this became broken.


You should not be able to have an empty password login, but it is clearly not being properly checked.
'You must enter your password!' error pops up when you do this with most accounts. However, when you do it on an account with no password that verification step is completely missed.

This is a bug, and really should be seen as one. 

If you are a server owner and have set empty passwords in the past, you can verify by testing one account. Go into your mysql database under sb_admins, and find the hash for that account. You can then bulk update your users with: update sb_admins SET password='aNewHash' WHERE password='blankhash';

**Comment by KoleBunny** [comment]
I've gone around and tested this on loads of sourceban servers, I don't think this is server admins fault. This is a bug that needs to be be fixed ASAP.

**Comment by RonMelkhior** [comment]
> This is a bug that needs to be be fixed ASAP.

Only thing that is a bug about it is that SourceBans allowed blank passwords.  
Again, people could have avoided it if they just set a password, it's not like having a Steam account linked means there is no password.

**Comment by KoleBunny** [comment]
With the new version, you just need a steam login. And like this server owner and most others, if an account had a password set to nothing, it was locked and just couldn't be accessed. A bug happened and now any account they had locked with the empty password, has nothing as a password .

**Comment by KoleBunny** [comment]
I want you to go to any sourcebans for any server, try to login as a random admin. And it works 99% of the time. It's not the server owners, its the plugin changing.
Oh and this guys site is getting "hacked" at the moment. 
![lmao](https://cloud.githubusercontent.com/assets/11532552/22451179/334752e2-e73a-11e6-901f-fd9a517c5013.png)

**Comment by eNNercY1337** [comment]
Not everyone wants a steam only login and neither do i.

As mentioned already several times its pure stupidity if you leave an required field empty 😩

**Comment by KoleBunny** [comment]
I don't know if i'm not being clear. Or if the server owner explaining isn't being clear either. Leaving the password field blank would LOCK the account, you couldn't log in with the name. For some reason theres a bug where now all accounts locked that way, have no password. Giving anyone in any sourceban webpage access to an admin account. 
This is not the server owners fault. Check out any sourceban page you want, login as random admins. Will work 99% of the time.

**Comment by KoleBunny** [comment]
I'm standing behind that this is a bug that needs to be fixed ASAP, if so many server admins ran accounts with no passwords, wouldn't this be a big deal sooner? 
I think this is a very scary thing and is going to effect a lot of server owners who're asleep and can't update or enter the mysql database fix that Muny posted.

**Comment by Munzy** [comment]
This isn't a required field. You can set any account up without a password, with zero issue. It doesn't even stop you. The only required fields are steamid, name, and email. 

Clearly there is a further bug that you are sanitizing passwords being entered into the login page after the fact.

**Comment by ciszard** [comment]
@KoleBunny - how is that screen shot an example of a site being "hacked"

The errors you see there are attributed to deprecated PHP calls in PHP 7.0 (the site you linked was running PHP 7, not supported by sbans) - nothing to do with this bug, or any authentication module. Also the site you included a screen shot of is not using sourcebans-pp.

**Comment by KoleBunny** [comment]
Epod, I'm not the best at what i'm talking about, to me it looked like someone was trying to XSS into the site through that site, but yes I fucked up i'm being very impulsive and shouldn't jump tht fast to conclusion, there is still 100% the login bug.

**Comment by Munzy** [comment]
For example, if you normally try a blank password this error comes up: 

![Login with no password](https://www.cameronmunroe.com/u/explorer_2017-01-30_20-16-49.png)

However, when you enter a blank password for an account that has no setup password it skips that check entirely.

**Comment by KoleBunny** [comment]
@Munroenet It says "You must enter a password" for me when you watch the login videos. But still passed through.

**Comment by Munzy** [comment]
The CheckLogin function clearly should return false; on the password and fail it, but it isn't.

**Comment by KoleBunny** [comment]
Just a heads up since people have threatened to get me arrested for tampering with their servers.
This is perfectly legal as I have not destroyed any files, and did not use anything to tamper into the server. 
Thank you.

**Comment by ciszard** [comment]
@KoleBunny  - in regards to your deleted comment about server owners threatening you with jail time. While we were not one of the server's that threatened you, you did come into our teamspeak and reported the bug (which we are unaffected by) in a very incriminating manner, regardless if that was your intention or not. 

As someone who works in the security field, I highly recommend you familiarize yourself with responsible disclosure policy's that some organizations put out, and apply their practices to your future vulnerability disclosures. 

In terms of the legality of the issue, I doubt any one of us in this threat is a lawyer and can make such claims. However remember this is America and anyone and sue anyone for anything. Have a look at https://www.techdirt.com/articles/20130318/23033422370/expose-blatant-security-hole-ats-servers-get-35-years-jail.shtml

While I can see how you are trying to help in a way you think is best, I urge you to consider reevaluating how you go about this in the future - especially if you plan on potentially disclosing vulnerabilitys to actual organizations. 

Feel free to find me if you would like to talk about this subject.

**Comment by KoleBunny** [comment]
The reason why I was fine with disclosing this bug one of my friends (Sys admin for microsoft) told me since these logins aren't protected, and i'm not destroying anything it's perfectly legal for me to do so. 

Reason why I deleted the threaten thing because I realize it's just some kids who have nothing better to do.
I felt like I needed to disclose this to as many people who had web admin accounts that anyone could login to, and seeing as this a (new?) bug I thought it would be good to inform people to protect their shit.
